transport.go

```diff
@@ -202,6 +224,49 @@ func (t *transport) Dial(ctx context.Context, raddr ma.Multiaddr, p peer.ID) (tp
 	return conn, nil
 }
 
+func (t *transport) holePunch(ctx context.Context, network string, addr *net.UDPAddr) (tpt.CapableConn, error) {
+	pconn, err := t.connManager.Dial(network, addr)
+	if err != nil {
+		return nil, err
+	}
+	defer pconn.DecreaseCount()
+
+	ctx, cancel := context.WithTimeout(ctx, 5*time.Second)
+	defer cancel()
+
+	connCh := make(chan tpt.CapableConn)
+
+	key := addr.String()
+	t.holePunchingMx.Lock()
+	t.holePunching[key] = activeHolePunch{connCh: connCh, ctx: ctx}
+	t.holePunchingMx.Unlock()
+
+	payload := make([]byte, 64)
+	for i := 0; ; i++ {
+		if _, err := rand.Read(payload); err != nil {
+			return nil, err
+		}
+		if _, err := pconn.UDPConn.WriteToUDP(payload, addr); err != nil {
+			return nil, err
+		}
+
+		maxSleep := 10 * (i + 1) * (i + 1) // in ms
+		if maxSleep > 200 {
+			maxSleep = 200
+		}
+		select {
+		case c := <-connCh:
+			return c, nil
+		case <-time.After(10*time.Millisecond + time.Duration(rand.Intn(maxSleep))*time.Millisecond):
+		case <-ctx.Done():
+			t.holePunchingMx.Lock()
+			delete(t.holePunching, key)
+			t.holePunchingMx.Unlock()
```

Comment:
Check to see if _we_ succeeded in deleting it. If not, we _know_ that we're about to receive the conn, so take it. This ensures that we can't drop a connection.